File: 03/dotnet6/binary.cs
int[] Test = { 0b00100, 0b11110, 0b10110, 0b10111, 0b10101, 0b01111, 0b00111, 0b11100, 0b10000, 0b11001, 0b00010, 0b01010 };

var path = System.Reflection.Assembly.GetExecutingAssembly().Location;
var temp = System.IO.File.ReadAllLines($"{Path.GetDirectoryName(path)}\\input.txt");
var Real = (from x in temp
            select Convert.ToInt32(x, 2)).ToArray();
Console.WriteLine($"test = {Day3(Test)}");
Console.WriteLine($"part1 = {Day3(Real)}");
Console.WriteLine($"part2 test = {Day3Pt2(Test)}");
Console.WriteLine($"part2 = {Day3Pt2(Real)}");

static int Day3(int[] readings)
{
    var bits = Convert.ToString(readings.Max(), 2).Length - 1;
    int gamma = 0, epsilon = 0;
    for (var x = bits; x >= 0; x--)
    {
        var bit = 1 << x;
        int ones = 0, zeros = 0;
        foreach (var y in readings)
        {
            if ((y & bit) > 0)
            {
                ones++;
            }
            else
            {
                zeros++;
            }
        }
        gamma += ones > zeros ? bit : 0;
        epsilon += ones > zeros ? 0 : bit;
    }
    return gamma * epsilon;
}

static int Day3Pt2(int[] readings)
{
    var list = readings.ToList();
    var bits = Convert.ToString(readings.Max(), 2).Length - 1;
    for (var bitNumber = bits; bitNumber >= 0; bitNumber--)
    {
        var bit = 1 << bitNumber;
        int ones = 0, zeros = 0;
        foreach (var y in list)
        {
            if ((y & bit) > 0)
            {
                ones++;
            }
            else
            {
                zeros++;
            }
        }
        var toRemove = ones < zeros ? bit : 0;

        var list1 = list.Where(x => (x & bit) == toRemove);
        list.RemoveAll(x => (list1.Contains(x)));
        list.TrimExcess();
        if (list.Count == 1)
        {
            break;
        }
    }
    var gamma = list[0];
    list = readings.ToList();
    for (var bitNumber = bits; bitNumber >= 0; bitNumber--)
    {
        var bit = 1 << bitNumber;
        int ones = 0, zeros = 0;
        foreach (var y in list)
        {
            if ((y & bit) > 0)
            {
                ones++;
            }
            else
            {
                zeros++;
            }
        }
        var toRemove = ones < zeros ? bit : 0;

        var list1 = list.Where(x => (x & bit) == toRemove);
        list.RemoveAll(x => (!list1.Contains(x)));
        list.TrimExcess();

        if (list.Count == 1)
        {
            break;
        }
    }
    var epsilon = list[0];
    return gamma * epsilon;
}

File: 03/python/binary.py
import os
import sys


def part1(readings):
    bits = len(bin(max(readings))) - 2
    gamma = 0
    epsilon = 0
    for x in range(bits - 1, -1, -1):
        bit = 1 << x
        ones = 0
        zeros = 0
        for y in readings:
            if y & bit > 0:
                ones += 1
            else:
                zeros += 1
        if ones > zeros:
            gamma += bit
        else:
            epsilon += bit
    return gamma * epsilon


def part2(readings):

    # Part 2
    bits = len(bin(max(readings))) - 2
    temp_list = []
    for item in readings:
        temp_list.append(item)
    list1 = []
    for x in range(bits - 1, -1, -1):
        bit = 1 << x
        ones = 0
        zeros = 0
        for y in temp_list:
            if y & bit > 0:
                ones += 1
            else:
                zeros += 1
        if ones < zeros:
            to_remove = bit
        else:
            to_remove = 0
        list1.clear()
        for reading in temp_list:
            if reading & bit != to_remove:
                list1.append(reading)
        temp_list.clear()
        for item in list1:
            temp_list.append(item)
        if len(temp_list) == 1:
            break
    gamma = temp_list[0]

    temp_list.clear()
    for item in readings:
        temp_list.append(item)
    for x in range(bits - 1, -1, -1):
        bit = 1 << x
        ones = 0
        zeros = 0
        for y in temp_list:
            if y & bit > 0:
                ones += 1
            else:
                zeros += 1
        if ones < zeros:
            to_remove = bit
        else:
            to_remove = 0
        list1.clear()
        for reading in temp_list:
            if reading & bit == to_remove:
                list1.append(reading)
        temp_list.clear()
        for item in list1:
            temp_list.append(item)
        if len(temp_list) == 1:
            break
    epsilon = temp_list[0]  # should be 10 in test
    return gamma * epsilon


def main():
    file_path = os.path.join(os.path.dirname(__file__), "input.txt")
    with open(file_path, "+rt") as f:
        lines = f.readlines()
        int_list = [int(i, 2) for i in lines]
    print("Part 1 %d" % (part1(int_list)))
    print("Part 2 %d" % (part2(int_list)))


if __name__ == "__main__":
    sys.exit(int(main() or 0))
